File: sdk/csharp/src/Providers/Apihz.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// apihz（接口盒子）渠道 — https://apihz.cn
/// 需 id/key 凭据，内置官方公共账号 88888888/88888888，可用 APIHZ_ID/APIHZ_KEY 覆盖。
/// 创建：GET /api/mail/mailcache.php 建邮箱（有效期 10 分钟）。
/// 收信：GET /api/mail/mailgetlist.php（须带创建时的 pwd）。
/// token 编码 "apihz1:" + base64(JSON{mail,pwd})。
/// </summary>
public static class Apihz
{
    private const string BaseUrl = "https://cn.apihz.cn";
    private const string TokPrefix = "apihz1:";
    private const string PublicId = "88888888";
    private const string PublicKey = "88888888";

    // apihz 自研收信域名
    private static readonly string[] Domains = { "apimail.email", "apimail.vip" };
    private static readonly Random Rand = new();

    private static string Ua()
    {
        var cfg = Config.Get();
        if (cfg.Headers is not null && cfg.Headers.TryGetValue("User-Agent", out var ua) && !string.IsNullOrEmpty(ua))
            return ua;
        return "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";
    }

    private static (string id, string key) Credentials()
    {
        var cfg = Config.Get();
        var id = string.IsNullOrWhiteSpace(cfg.ApihzId) ? PublicId : cfg.ApihzId!.Trim();
        var key = string.IsNullOrWhiteSpace(cfg.ApihzKey) ? PublicKey : cfg.ApihzKey!.Trim();
        return (id, key);
    }

    private static string RandomLocal(int n)
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new StringBuilder();
        for (var i = 0; i < n; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static string RandomPassword()
    {
        const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new StringBuilder();
        for (var i = 0; i < 12; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static string EncToken(string mail, string pwd)
    {
        var raw = Json.Serialize(new Dictionary<string, object?> { ["mail"] = mail, ["pwd"] = pwd });
        return TokPrefix + Convert.ToBase64String(Encoding.UTF8.GetBytes(raw));
    }

    private static (string mail, string pwd) DecToken(string? tok)
    {
        if (string.IsNullOrEmpty(tok) || !tok.StartsWith(TokPrefix, StringComparison.Ordinal))
            throw new Exception("apihz: 无效 token");
        string mail, pwd;
        try
        {
            var raw = Encoding.UTF8.GetString(Convert.FromBase64String(tok[TokPrefix.Length..]));
            var obj = Json.Parse(raw) as JsonObject ?? throw new Exception();
            mail = Json.Str(obj, "mail").Trim();
            pwd = Json.Str(obj, "pwd").Trim();
        }
        catch { throw new Exception("apihz: 无效 token"); }
        if (mail.Length == 0 || pwd.Length == 0) throw new Exception("apihz: 无效 token");
        return (mail, pwd);
    }

    private static string First(JsonObject msg, params string[] keys)
    {
        foreach (var k in keys)
        {
            var v = Json.Str(msg, k);
            if (!string.IsNullOrEmpty(v)) return v;
        }
        return "";
    }

    /// <summary>创建 apihz 临时邮箱，有效期 10 分钟</summary>
    public static EmailInfo Generate()
    {
        var (id, key) = Credentials();
        var domain = Domains[Rand.Next(Domains.Length)];
        var name = RandomLocal(10);
        var pwd = RandomPassword();

        var url = $"{BaseUrl}/api/mail/mailcache.php?id={WebUtility.UrlEncode(id)}&key={WebUtility.UrlEncode(key)}" +
                  $"&domain={WebUtility.UrlEncode(domain)}&name={WebUtility.UrlEncode(name)}&pwd={WebUtility.UrlEncode(pwd)}&buytype=0";
        var resp = Http.Get(url, new Dictionary<string, string> { ["User-Agent"] = Ua(), ["Accept"] = "application/json" });
        resp.EnsureSuccess();

        var data = Json.Parse(resp.Body) as JsonObject;
        var code = data?["code"]?.GetValue<int>() ?? 0;
        var mail = Json.Str(data, "mail");
        if (data is null || code != 200 || string.IsNullOrEmpty(mail))
            throw new Exception($"apihz: 创建邮箱失败 {resp.Body}");

        // 优先使用响应回传的 pwd，确保读信密码正确
        var finalPwd = Json.Str(data, "pwd");
        if (string.IsNullOrEmpty(finalPwd)) finalPwd = pwd;

        return new EmailInfo("apihz", mail, EncToken(mail, finalPwd));
    }

    /// <summary>获取 apihz 邮件列表（须带创建时的 mail 与 pwd）</summary>
    public static List<Email> GetEmails(string? token)
    {
        var (mail, pwd) = DecToken(token);
        var (id, key) = Credentials();

        var url = $"{BaseUrl}/api/mail/mailgetlist.php?id={WebUtility.UrlEncode(id)}&key={WebUtility.UrlEncode(key)}" +
                  $"&mail={WebUtility.UrlEncode(mail)}&pwd={WebUtility.UrlEncode(pwd)}&page=1";
        var resp = Http.Get(url, new Dictionary<string, string> { ["User-Agent"] = Ua(), ["Accept"] = "application/json" });
        resp.EnsureSuccess();

        var data = Json.Parse(resp.Body) as JsonObject;
        var code = data?["code"]?.GetValue<int>() ?? 0;
        var result = new List<Email>();
        if (data is null || code != 200) return result;

        if (data["data"] is not JsonArray items) return result;
        foreach (var node in items)
        {
            if (node is not JsonObject msg) continue;
            result.Add(Normalize.NormalizeEmail(new Dictionary<string, object?>
            {
                ["id"] = First(msg, "time1"),
                ["from"] = First(msg, "frommail", "fromname"),
                ["to"] = mail,
                ["subject"] = First(msg, "subject"),
                ["text"] = First(msg, "content"),
                ["html"] = First(msg, "content"),
                ["date"] = First(msg, "time2", "time1"),
            }, mail));
        }
        return result;
    }
}

File: sdk/php/src/Providers/Apihz.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\ConfigStore;
use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * apihz（接口盒子）渠道实现 — https://apihz.cn
 *
 * 需 id/key 凭据，内置公共账号 88888888/88888888（共享频次），
 * 可用配置 apihzId/apihzKey 或环境变量 APIHZ_ID/APIHZ_KEY 覆盖。
 *   GET /api/mail/mailcache.php 创建邮箱（有效期 10 分钟）；
 *   GET /api/mail/mailgetlist.php 读信（须携带创建时的 pwd）。
 * token 编码: "apihz1:" + base64(JSON{mail,pwd})。
 */
final class Apihz
{
    private const BASE_URL = 'https://cn.apihz.cn';
    private const TOK_PREFIX = 'apihz1:';
    private const PUBLIC_ID = '88888888';
    private const PUBLIC_KEY = '88888888';

    /** @var string[] */
    private const DOMAINS = ['apimail.email', 'apimail.vip'];

    private const DEFAULT_UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36';

    private static function userAgent(): string
    {
        foreach (ConfigStore::get()->headers as $k => $v) {
            if (strtolower((string) $k) === 'user-agent' && $v !== '') {
                return (string) $v;
            }
        }
        return self::DEFAULT_UA;
    }

    /**
     * 读取 apihz 调用凭据：优先配置，回退官方公共账号。
     *
     * @return array{0:string,1:string}
     */
    private static function credentials(): array
    {
        $cfg = ConfigStore::get();
        $cid = trim((string) ($cfg->apihzId ?? '')) ?: self::PUBLIC_ID;
        $ckey = trim((string) ($cfg->apihzKey ?? '')) ?: self::PUBLIC_KEY;
        return [$cid, $ckey];
    }

    private static function randomLocal(int $n): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = '';
        for ($i = 0; $i < $n; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    private static function randomPassword(): string
    {
        $chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        $out = '';
        for ($i = 0; $i < 12; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    private static function encToken(string $mail, string $pwd): string
    {
        $raw = json_encode(['mail' => $mail, 'pwd' => $pwd], JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
        return self::TOK_PREFIX . base64_encode((string) $raw);
    }

    /**
     * @return array{0:string,1:string}
     */
    private static function decToken(?string $tok): array
    {
        if ($tok === null || !str_starts_with($tok, self::TOK_PREFIX)) {
            throw new \InvalidArgumentException('apihz: 无效 token');
        }
        $decoded = base64_decode(substr($tok, strlen(self::TOK_PREFIX)), true);
        if ($decoded === false) {
            throw new \InvalidArgumentException('apihz: 无效 token');
        }
        $obj = json_decode($decoded, true);
        if (!is_array($obj)) {
            throw new \InvalidArgumentException('apihz: 无效 token');
        }
        $mail = trim((string) ($obj['mail'] ?? ''));
        $pwd = trim((string) ($obj['pwd'] ?? ''));
        if ($mail === '' || $pwd === '') {
            throw new \InvalidArgumentException('apihz: 无效 token');
        }
        return [$mail, $pwd];
    }

    /**
     * @param array<mixed> $msg
     * @param string[]     $keys
     */
    private static function first(array $msg, array $keys): string
    {
        foreach ($keys as $key) {
            $val = $msg[$key] ?? null;
            if ($val !== null && $val !== '') {
                return (string) $val;
            }
        }
        return '';
    }

    public static function generate(): EmailInfo
    {
        [$cid, $ckey] = self::credentials();
        $domain = self::DOMAINS[array_rand(self::DOMAINS)];
        $name = self::randomLocal(10);
        $pwd = self::randomPassword();

        $resp = HttpClient::get(
            self::BASE_URL . '/api/mail/mailcache.php',
            ['User-Agent' => self::userAgent(), 'Accept' => 'application/json'],
            query: [
                'id' => $cid,
                'key' => $ckey,
                'domain' => $domain,
                'name' => $name,
                'pwd' => $pwd,
                'buytype' => '0',
            ],
        );
        $data = HttpClient::json($resp);
        if (($data['code'] ?? null) !== 200 || empty($data['mail'])) {
            throw new \RuntimeException('apihz: 创建邮箱失败');
        }
        $mail = (string) $data['mail'];
        $finalPwd = (string) ($data['pwd'] ?? $pwd);
        return new EmailInfo('apihz', $mail, self::encToken($mail, $finalPwd));
    }

    /**
     * @return Email[]
     */
    public static function getEmails(?string $token): array
    {
        [$mail, $pwd] = self::decToken($token);
        [$cid, $ckey] = self::credentials();

        $resp = HttpClient::get(
            self::BASE_URL . '/api/mail/mailgetlist.php',
            ['User-Agent' => self::userAgent(), 'Accept' => 'application/json'],
            query: ['id' => $cid, 'key' => $ckey, 'mail' => $mail, 'pwd' => $pwd, 'page' => '1'],
        );
        $data = HttpClient::json($resp);
        if (($data['code'] ?? null) !== 200) {
            return [];
        }
        $items = $data['data'] ?? null;
        if (!is_array($items)) {
            return [];
        }

        $out = [];
        foreach ($items as $msg) {
            if (!is_array($msg)) {
                continue;
            }
            $raw = [
                'id' => self::first($msg, ['time1']),
                'from' => self::first($msg, ['frommail', 'fromname']),
                'to' => $mail,
                'subject' => self::first($msg, ['subject']),
                'text' => self::first($msg, ['content']),
                'html' => self::first($msg, ['content']),
                'date' => self::first($msg, ['time2', 'time1']),
            ];
            $out[] = Normalize::email($raw, $mail);
        }
        return $out;
    }
}
